src/serverenvironment.cpp

```diff
@@ -360,33 +361,25 @@ void ActiveBlockList::update(std::vector<PlayerSAO*> &active_players,
 				playersao->getEyePosition(),
 				camera_dir,
 				playersao->getFov(),
-				newlist);
+				extralist);
 		}
 	}
 
-	/*
-		Find out which blocks on the old list are not on the new list
-	*/
-	// Go through old list
-	for (v3s16 p : m_list) {
-		// If not on new list, it's been removed
-		if (newlist.find(p) == newlist.end())
-			blocks_removed.insert(p);
-	}
-
-	/*
-		Find out which blocks on the new list are not on the old list
-	*/
-	// Go through new list
+	// make sure extra list only contains extra blocks
 	for (v3s16 p : newlist) {
-		// If not on old list, it's been added
-		if (m_list.find(p) == m_list.end())
-			blocks_added.insert(p);
+		extralist.erase(p);
 	}
 
-	/*
-		Update m_list
-	*/
+	// newlist now contains all blocks
+	newlist.insert(extralist.begin(), extralist.end());
+
+	// Find out which blocks on the old list are not on the new list
+	std::set_difference(m_list.begin(), m_list.end(), newlist.begin(), newlist.end(), std::inserter(blocks_removed, blocks_removed.end()));
+
+	// Find out which blocks on the new list are not on the old list
+	std::set_difference(newlist.begin(), newlist.end(), m_list.begin(), m_list.end(), std::inserter(blocks_added, blocks_added.end()));
```

Comment:
Wait, no, this would again produce the added set with all added block, including the the extra set... Fix coming momentarily.

Reply:
I think I like the first version better now... At least there it was clear.